testing.rst

```diff
@@ -11,377 +11,513 @@ using both functional and unit tests.
+Tests that interact with the database should use their own separate
+database to not mess with the databases used in the other
+:ref:`configuration environments <configuration-environments>`.
 
-In the above example, you validated that the HTTP response was successful. The
-next step is to validate that the page actually contains the expected content.
-The ``createClient()`` method returns a client, which is like a browser that
-you'll use to crawl your site::
+To do that, edit or create the ``.env.test.local`` file at the root
+directory of your project and define the new value for the ``DATABASE_URL``
+env var:
 
-    $crawler = $client->request('GET', '/post/hello-world');
+.. code-block:: env
 
-The ``request()`` method (read
-:ref:`more about the request method <testing-request-method-sidebar>`)
-returns a :class:`Symfony\\Component\\DomCrawler\\Crawler` object which can
-be used to select elements in the response, click on links and submit forms.
+    # .env.test.local
+    DATABASE_URL="mysql://USERNAME:PASSWORD@127.0.0.1:3306/DB_NAME?serverVersion=5.7"
+
+.. TODO Symfony CLI automatically appends `_test` somehow sometimes, right?
+
+This assumes that each developer/machine uses a different database for the
+tests. If the test set-up is the same on each machine, use the ``.env.test``
+file instead and commit it to the shared repository. Learn more about
+:ref:`using multiple .env files in Symfony applications <configuration-multiple-env-files>`.
+
+After that, you can create the test database and all tables using:
+
+.. code-block:: terminal
+
+    # create the test database
+    $ php bin/console --env=test doctrine:database:create
+
+    # create the tables/columns in the test database
+    $ php bin/console --env=test doctrine:schema:create
 
 .. tip::
 
-    The ``Crawler`` only works when the response is an XML or an HTML document.
-    To get the raw content response, call ``$client->getResponse()->getContent()``.
+    A common practice is to append the ``_test`` suffix to the original
+    database names in tests. If the database name in production is called
+    ``project_acme`` the name of the testing database could be
+    ``project_acme_test``.
+
+Resetting the Database Automatically Before each Test
+.....................................................
+
+Tests should be independent from each other to avoid side effects. For
+example, if some test modifies the database (by adding or removing an
+entity) it could change the results of other tests.
 
-The crawler integrates with the ``symfony/css-selector`` component to give you the
-power of CSS selectors to find content in a page. To install the CSS selector
-component, run:
+The `DAMADoctrineTestBundle`_ uses Doctrine transactions to let each test
+interact with an unmodified database. Install it using:
 
 .. code-block:: terminal
 
-    $ composer require --dev symfony/css-selector
+    $ composer require --dev dama/doctrine-test-bundle
```

Comment:
Shouldn't we add this package to `test-pack` then?

Reply:
I agree. However, as we know, there is no official reset tool for Doctrine, nor is there a library that only does the reset logic. So we'll have to wait for Foundry to be documented in this article before we can recommend the `ResetDatabase` trait for resetting and only use the dama bundle for performance.

Also, this PR targets 4.4 (to avoid merge conflicts for the next years). We'll probably only want to talk about Foundry in the 5.3+ version of the docs.